Add Mesh class

sbrus89 opened the pull request:
<!--
Please add a description of what is accomplished in the PR here at the top:
-->

This PR introduces the Mesh class for the local mesh variables on an MPI rank. It includes unit tests that check mesh variables against various metrics to ensure the values have been read correctly.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] Documentation:
  * [x] Design document has been generated and added to the docs
  * [x] User's Guide has been updated
  * [x] Developer's Guide has been updated
  * [x] Documentation has been [built locally](https://e3sm-project.github.io/Omega/develop/devGuide/BuildDocs.html) and changes look as expected
* [x] Testing
  * [x] A comment in the PR documents testing used to verify the changes including any tests that are added/modified/impacted.
  * [x] CTest unit tests for new features have been added per the approved design. 
  * [x] Unit tests have passed. Please provide a relevant CDash build entry for verification.
after.

<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

sbrus89 commented:
I still need to add documentation, but have opened a PR since the code portion is nearly complete.

mwarusz commented on components/omega/src/ocn/HorzMesh.h:
```diff
@@ -0,0 +1,192 @@
+
+   // Mesh connectivity
+
+   Array2DI4 CellsOnCell;      ///< Indx of cells that neighbor each cell
+   ArrayHost2DI4 CellsOnCellH; ///< Indx of cells that neighbor each cell
```
This pattern of declaring two arrays (host and device) per mesh variable will double the amount of memory needed to store the mesh when the device is a CPU. Additionally, redundant copies between the two arrays will be performed. It is probably fine for now, but long-term we might need a different solution.

brian-oneill commented on components/omega/test/CMakeLists.txt:
```diff
@@ -160,6 +160,42 @@ add_test(
+  ${OMEGA_SOURCE_DIR}/src/base
+  ${OMEGA_SOURCE_DIR}/src/infra
+  ${OMEGA_SOURCE_DIR}/src/ocn
+  ${PARMETIS_ROOT}/include
```
This was changed in #44, should be `${Parmetis_INCLUDE_DIRS}`

sbrus89 replied:
Thanks @brian-oneill!

brian-oneill commented on components/omega/test/CMakeLists.txt:
`HORZMESH_TEST` should be added to the list in `set_tests_properties` at the bottom of this file so ctest can catch FAILs.

brian-oneill reviewed (changes requested):
Still reviewing the code, but with the needed change to the parmetis location variable in the CMake configuration, I built and successfully ran the test on Chrysalis and Frontier.

mwarusz commented:
Just noting that for implementing differential operators it would be good for the `HorzMesh` class to compute and store `EdgeSignOnCell` and `EdgeSignOnVertex`.

sbrus89 commented:
thanks @mwarusz, yes I realized that an am adding computations.

philipwjones commented on components/omega/test/CMakeLists.txt:
```diff
@@ -160,6 +160,42 @@ add_test(
+add_test(
+  NAME HORZMESH_TEST
+  COMMAND ${MPI_EXEC} -n 8 -- ./${_TestHorzMeshName}
+)
```
You will also need to add HORZMESH_TEST to the set_test_properties list at the end of this file

philipwjones reviewed (changes requested):
This looks mostly good.  A few questions/comments and the specific change requested in a separate comment.

Any reason not to read all the quantities at once with a single read routine? That way we don't need to retain the IO decompositions and other IO stuff in the class.

It would be good to flesh out a little more detail in the DevGuide how you would actually use this class throughout the code.

Given that the instance will be created in an Omega init routine, how would you access this by the run method? We won't be able to pass the instance up through the coupler. This is one reason the other classes have been storing all instances for later retrieval. Another option is to have some big omega header with extern static instance for stuff that needs to be saved from one phase to another.

sbrus89 commented:
@philipwjones, thanks for the comments. I'll make edits to have a single read routine and instance storage/retrieval.

sbrus89 commented:
@philipwjones, my thought with the multiple read routines was that 1.) it breaks up what is otherwise a pretty long routine into more coherent pieces and 2.) it would make it easier to gradually transition to computing (vs reading) the dependent mesh quantities as we had discussed. I could add a routine to clean up all the IO decomps after the read routines. 

That said, I also don't mind making this into a single routine.

philipwjones commented:
@sbrus89 Thanks for the explanation - I guess we can leave the multiple reads for now and consolidate later once we've determined what we're reading vs computing.

sbrus89 commented:
Thanks again for your comments @philipwjones, @brian-oneill, and @mwarusz. I think I have addressed the the changes suggested so far. Please let me know if anything else comes to mind.

sbrus89 commented:
I realized some build issues, so I'm sorting those out.

mwarusz commented on components/omega/src/ocn/HorzMesh.h:
```diff
@@ -0,0 +1,231 @@
+   ~HorzMesh();
+
+   /// Deallocates arrays
+   void clear();
+
+   /// Remove mesh by name
+   void erase(std::string InName);
```
I think that `clear` and `erase` should be static methods as in `Decomp`.

sbrus89 replied:
thanks, @mwarusz. I just corrected that.

philipwjones commented:
@sbrus89 Just a couple (hopefully final) things.  First, Omega is following the MPAS convention of adding boundary conditions in the NCellsAll+1 location so all arrays should be dimensioned NCellsSize (and NEdgesSize, etc) to accommodate the extra entry since the connectivity arrays will point to that location for boundary Cells/Edges/Vertices.  Second, can you verify that the halo cells are filled correctly?  Since the CellID, etc arrays do have the IDs in the halo regions, the parallel IO may be filling them correctly, but I have not checked that - you could do a simple before/after Halo update comparison to see.

Built the docs and they look fine and verified unit tests pass, but will recheck after the above changes.

sbrus89 commented:
@philipwjones, thanks for noticing that. I fixed the array sizes and added a test that verifies the read-in values for the halos based on CellID etc. are correct by checking them against a full halo exchange.

philipwjones reviewed (approved):
Thanks for the changes @sbrus89 . This looks fine to me now and still passes unit tests.

mark-petersen commented:
I rebased locally on the `develop` branch because master was just merged in. There were no conflicts. I tested as follows

<details>
<summary> test instructions: </summary> <p>

```

######### Frontier
cd /ccs/home/mpetersen/repos/omega/pr
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd components/omega/
module load cmake
rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=$PROJWORK/cli115/pwjones/frontierlibs-cray/parmetis
cmake \
   -DOMEGA_CIME_COMPILER=crayclang \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S .. -B .
./omega_build.sh
ln -isf /ccs/home/mpetersen/meshes/ocean.QU.240km.151209.nc test/OmegaMesh.nc

salloc -A cli115 -J inter -t 2:00:00 -q debug -N 1 -S 0
cd /ccs/home/mpetersen/repos/omega/pr/components/omega/build
./omega_ctest.sh

######### perlmutter
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd components/omega/
module load cmake
mkdir build
cd build
export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_3_0_gnu_mpich/var/spack/environments/dev_polaris_0_3_0_gnu_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=intel \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S .. -B .
ln -s /global/cfs/cdirs/e3sm/inputdata/ocn/mpas-o/oQU240/ocean.QU.240km.151209.nc test/OmegaMesh.nc
./omega_build.sh

salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=e3sm
./omega_ctest.sh

######### chrysalis
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd components/omega/
module load cmake
mkdir build
cd build
export PARMETIS_ROOT=/lcrc/group/e3sm/ac.boneill/intel-openmpi-libs
cmake \
   -DOMEGA_CIME_COMPILER=intel \
   -DOMEGA_CIME_MACHINE=chrysalis \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S .. -B .
ln -isf /home/ac.mpetersen/inputdata/ocn/mpas-o/oQU240/ocean.QU.240km.151209.nc test/OmegaMesh.nc
./omega_build.sh

srun -p debug -N 1 -t 1:00:00 --pty bash
./omega_ctest.sh
```
</details>

1. **chrysalis**: All tests pass, including the new `HORZMESH_TEST`. 
2. **frontier**: All tests pass except `IO_TEST` due to a memory high water mark. That is unrelated to this PR, but I show it here for reference:

<details>
<summary> error: </summary> <p>

```

Test project /ccs/home/mpetersen/repos/omega/pr/components/omega/build
    Start 8: IO_TEST
1/1 Test #8: IO_TEST ..........................***Failed    3.50 sec
Using memory pool. Initial size: 62.8379GB ;  Grow size: 62.8379GB.
[2024-02-05 13:52:28.894] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.895] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.895] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.896] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.897] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.898] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.899] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
[2024-02-05 13:52:28.899] [info] [IOTest.cpp:76] IOTest: Default decomp retrieval PASS
PIO: FATAL ERROR: Aborting... FATAL ERROR: Permission denied (file = IOTest.nc) (/ccs/home/mpetersen/repos/omega/pr/externals/scorpio/src/clib/pioc_support.c: 3429)
Obtained 9 stack frames.
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x387aa4]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x387c92]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x388368]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x38e395]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x385f12]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x2f9448]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x2a7448]
/lib64/libc.so.6(__libc_start_main+0xef) [0x7fffe8baf29d]
/autofs/nccs-svm1_home1/mpetersen/repos/omega/pr/components/omega/build/test/./testIO.exe() [0x2a4d0a]
MPICH ERROR [Rank 0] [job id 1631852.8] [Mon Feb  5 13:52:29 2024] [frontier08681] - Abort(-1) (rank 0 in comm 0): application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0

aborting job:
application called MPI_Abort(MPI_COMM_WORLD, -1) - process 0
Pool Memory High Water Mark:       445952
Pool Memory High Water Efficiency: 6.60947e-06
ERROR: For the pool allocator labeled " �`




AKL's primary memory pool":
ERROR: Trying to free an invalid pointer
This means you have either already freed the pointer, or its address has been corrupted somehow.
terminate called after throwing an instance of 'std::runtime_error'
  what():  This means you have either already freed the pointer, or its address has been corrupted somehow.
srun: error: frontier08681: task 0: Aborted
srun: Terminating StepId=1631852.8
slurmstepd: error: *** STEP 1631852.8 ON frontier08681 CANCELLED AT 2024-02-05T13:52:31 ***
srun: error: frontier08681: tasks 1-3: Terminated
srun: error: frontier08688: tasks 4-7: Terminated
srun: Force Terminated StepId=1631852.8
```
</details>

3. **perlmutter**: I can't get cmake to work for this PR or for `develop`, so also unrelated to this PR.

<details>
<summary> error: </summary> <p>

```
cmake \
   -DOMEGA_CIME_COMPILER=intel \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S .. -B .

...

-- ===== Configuring GPTL library... =====
-- Could NOT find MPI_C (missing: MPI_C_WORKS)
-- Could NOT find MPI_Fortran (missing: MPI_Fortran_WORKS)
CMake Error at /global/common/software/nersc/pm-2021q4/sw/cmake-3.22.0/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
  Could NOT find MPI (missing: MPI_C_FOUND MPI_Fortran_FOUND C Fortran)
Call Stack (most recent call first):
  /global/common/software/nersc/pm-2021q4/sw/cmake-3.22.0/share/cmake-3.22/Modules/FindPackageHandleStandardArgs.cmake:594 (_FPHSA_FAILURE_MESSAGE)
  /global/common/software/nersc/pm-2021q4/sw/cmake-3.22.0/share/cmake-3.22/Modules/FindMPI.cmake:1830 (find_package_handle_standard_args)
  /global/homes/m/mpeterse/repos/omega/pr/externals/scorpio/src/gptl/CMakeLists.txt:140 (find_package)
```
</details>

philipwjones commented:
@mark-petersen The error you're getting on Frontier is because IOTest is trying to write a test output file and doesn't have permission. You'll have to run this from the /lustre/orion/scratch spaces on Frontier (eg ${MEMBERWORK}/cli115 )

mark-petersen commented:
Thanks @philipwjones and @grnydawn for the help. I made two mistakes. On perlmutter and frontier, I have to run tests on the scratch file system for PIO. On perlmutter, I had to `rm -rf build` directory to start a clean build and avoid that cmake error. Everything passes now. Here are my notes for reference.

<details>
<summary> instructions (change to your paths) </summary> <p>

```

######### Frontier
cd /ccs/home/mpetersen/repos/omega/pr
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd /lustre/orion/cli115/scratch/mpetersen/runs/240205_omega
module load cmake
rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=$PROJWORK/cli115/pwjones/frontierlibs-cray/parmetis
cmake \
   -DOMEGA_CIME_COMPILER=crayclang \
   -DOMEGA_CIME_MACHINE=frontier \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /ccs/home/mpetersen/repos/omega/pr/components/omega -B .
./omega_build.sh
ln -isf /ccs/home/mpetersen/meshes/ocean.QU.240km.151209.nc test/OmegaMesh.nc

salloc -A cli115 -J inter -t 2:00:00 -q debug -N 1 -S 0
cd /lustre/orion/cli115/scratch/mpetersen/runs/240205_omega
./omega_ctest.sh

######### perlmutter
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd components/omega/
module load cmake
cd /pscratch/sd/m/mpeterse/runs/240205_omega
rm -rf build
mkdir build
cd build
export PARMETIS_ROOT=/global/cfs/cdirs/e3sm/software/polaris/pm-cpu/spack/dev_polaris_0_3_0_gnu_mpich/var/spack/environments/dev_polaris_0_3_0_gnu_mpich/.spack-env/view
cmake \
   -DOMEGA_CIME_COMPILER=gnu  \
   -DOMEGA_CIME_MACHINE=pm-cpu \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S /global/homes/m/mpeterse/repos/omega/pr/components/omega -B .
ln -s /global/cfs/cdirs/e3sm/inputdata/ocn/mpas-o/oQU240/ocean.QU.240km.151209.nc test/OmegaMesh.nc
./omega_build.sh

salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=e3sm
cd /pscratch/sd/m/mpeterse/runs/240205_omega
./omega_ctest.sh

######### chrysalis
git submodule update --init --recursive externals/YAKL externals/ekat externals/scorpio cime
cd components/omega/
module load cmake
mkdir build
cd build
export PARMETIS_ROOT=/lcrc/group/e3sm/ac.boneill/intel-openmpi-libs
cmake \
   -DOMEGA_CIME_COMPILER=intel \
   -DOMEGA_CIME_MACHINE=chrysalis \
   -DOMEGA_PARMETIS_ROOT=${PARMETIS_ROOT}\
   -DOMEGA_BUILD_TEST=ON \
   -Wno-dev \
   -S .. -B .
ln -isf /home/ac.mpetersen/inputdata/ocn/mpas-o/oQU240/ocean.QU.240km.151209.nc test/OmegaMesh.nc
./omega_build.sh

srun -p debug -N 1 -t 1:00:00 --pty bash
./omega_ctest.sh


```
</details>

mark-petersen reviewed (approved):
Approving based on testing and review by Phil and Bryan. Thanks!

sbrus89 commented:
@mwarusz, let me know if you have any other comments. Otherwise I'll plan on merging this soon.

mwarusz commented on components/omega/doc/devGuide/HorzMesh.md:
```diff
@@ -0,0 +1,53 @@
+connectivity information from Decomp so this information can be passed among the
+computational routines, alongside the other local mesh information.  It then
+creates several parallel I/O decompositions and reads in the remaining subdomain
+mesh information.  Finally, any mesh information needed on the device to a
+device YAKL array from the host.  These tasks are organized into several private
```
There is something missing in the sentence beginning with "Finally" ("is copied to" ?). It might be a good idea to explicitly mention in this guide that the coordinate arrays are not available on the device.

sbrus89 replied:
Thanks @mwarusz!

mwarusz reviewed (approved):
I have one small comment regarding the dev docs. but this should be an easy change and everything else looks good so I am approving it.
